<img src="images/logos.png" alt="Banner" width="300"/>

<div>
    <font color=#690027 markdown="1">
        <h1>Statistiek in Python 1: Absolute en relatieve frequentie</h1> 
    </font>
</div>

<div>
    <font color=#690027 markdown="1">
        <h2>1. Werken met datasets</h2> 
    </font>
</div>

<div>
    <font color=#690027 markdown="1">
        <h3>1.1. Pandas dataframes</h3> 
    </font>
</div>

Statistiek is de wetenschap die een verzamelingen van gegevens onderzoekt op een wiskundige manier. Zo een verzameling gegevens gaan we verder een **dataset** noemen. De `pandas` module voegt een hoop handige functionaliteit om met datasets te werken aan Python toe.

In [1]:
import pandas as pd  # importeer de "pandas" module

Als eerste laat `pandas` ons toe om een dataset in te lezen met Python. Onderstaand commando laadt de dataset `dataset_voorbeeld` in, die is opgeslagen op de locatie `data/dataset_voorbeeld.csv`, en slaat ze op in een variabele `voorbeeld`.

<div class="alert alert-block alert-warning"> 
In het notebook <i>0300_SpreidingsdiagramDataBestand</i> werd uitgelegd hoe we met een csv-bestand in python moeten werken.
</div>

In [2]:
voorbeeld = pd.read_csv("data/dataset_voorbeeld.csv")
voorbeeld

Kolom 1  Kolom 2
0       A       12
1       A       11
2       C        7
3       A        3
4       B       19
5       A        8
6       C        8
7       C       12
8       A       31

Een dataset wordt voorgesteld in een tabel. In de *pandas* module noemt men deze datastructuur een **dataframe**.

Een kolom komt overeen met een **steekproefvariabele**, zoals bijvoorbeeld: massa, snelheid, temperatuur, scores,...

In een rij staat wat we in de statistiek een **element** noemen. Elk element bevat per kolom een **gegeven**. Een gegeven kan in vele vormen voorkomen: gehele getallen, reële getallen, tekst, symbolen, tijdstippen,...

<div class="alert alert-box alert-success">
In de statistiek gebruiken we de term 'variabele' om de grootheid van de data die we uit een steekproef halen aan te duiden. Bijvoorbeeld als we op 5 momenten de snelheid van een auto meten. Dan is het resultaat 5 gegevens uit de variabele 'snelheid auto'.

    
Wanneer we programmeren spreken we echter ook van een 'variabele'. Namelijk als we iets willen opslaan in het geheugen van de computer. 
    
Om verwarring te vermijden gebruiken we in wat volgt de term <b>steekproefvariabele</b> als we de  als we de data in een kolom van een dataframe bedoelen.
</div>

<div>
    <font color=#690027 markdown="1">
        <h3>1.2. Grootte van een dataset</h3> 
    </font>
</div>

Het aantal elementen, of aantal rijen, in een dataset noemen we de **grootte** van een dataset. De grootte kan je opvragen met de `len` functie. Dat is één van de vele mogelijke operaties op een dataset.

In [3]:
voorbeeld_grootte = len(voorbeeld)  # Sla de grootte van de de dataset "voorbeeld" op in de variabele "voorbeeld_lengte"
voorbeeld_grootte                   # Print de grootte op het scherm

9

`pandas` laat ons toe om een kolom uit een dataset af te zonderen en er berekeningen op uit te voeren.

In [4]:
kolom2 = voorbeeld["Kolom 2"]  # Zonder Kolom 2 af en sla ze op in de variabele "kolom2"
kolom2                         # Print Kolom 2 op het scherm

0    12
1    11
2     7
3     3
4    19
5     8
6     8
7    12
8    31
Name: Kolom 2, dtype: int64

Op een kolom kunnen we ook allerlei operaties uitvoeren. De functie `sum` geeft de som van alle gegevens in de kolom terug.

In [5]:
kolom2_som = kolom2.sum()  # Sla de som van de gegevens uit Kolom 2 op in de variabele kolom2_som
kolom2_som                 # Print de som van de gegevens uit Kolom 2 op het scherm

111

We kunnen ook aritmetische operaties (+,-,x,÷) op alle gegevens in een kolom tegelijk uitvoeren.

In [6]:
kolom2_dubbel = 2*kolom2  # Verdubbel alle gegevens uit Kolom 2 en sla de nieuwe kolom op in "kolom2_dubbel"
kolom2_dubbel             # Print "kolom2_dubbel" op het scherm

0    24
1    22
2    14
3     6
4    38
5    16
6    16
7    24
8    62
Name: Kolom 2, dtype: int64

<div>
    <font color=#690027 markdown="1">
        <h3>1.3. Absolute frequentie</h3> 
    </font>
</div>

Verder behandelen we de functie `value_counts`. Die telt hoeveel keer elk gegeven in een bepaalde kolom voorkomt. Het resultaat is een **absolute frequentietabel**.

In [7]:
absolute_frequentietabel_kolom2 = kolom2.value_counts()  # Tel hoeveel keer elk gegeven in Kolom 2 voorkomt
absolute_frequentietabel_kolom2                          # Print "kolom2_count" op het scherm

12    2
8     2
11    1
7     1
3     1
19    1
31    1
Name: Kolom 2, dtype: int64

### Opdracht 1.1

Zonder Kolom 1 af uit de voorbeeld dataset en print het resultaat op het scherm.

A    5
C    3
B    1
Name: Kolom 1, dtype: int64

Hoeveel keer komt het element *C* voor?

Antwoord: 

### Voorbeeld 1.1

Hoeveel keer één bepaald element voorkomt in een dataset noemen we de **absolute frequentie**. We kunnen die aflezen uit de absolute freqentietabel. Of we kunnen ze er met Pyhon zelf uithalen. Daarvoor passen we deze keer de naam van een element toe als sleutel.

In [9]:
absolute_frequentie_C = absolute_frequentietabel_kolom1['C']  # Sla de absulute frequentie van "C" op in de variabele absolute_frequentie_C
absolute_frequentie_C  # Print absolute_frequentie_C op het scherm

3

### Opdracht 1.2

Print de absolute frequentie van *A* op het scherm.

5

<div>
    <font color=#690027 markdown="1">
        <h3>1.4. Relatieve frequentie</h3> 
    </font>
</div>

### Opdracht 1.3

We willen nu de **relatieve frequentie** van berekenen. Dat is de verhouding tussen de absolute frequentie van een gegeven en de grootte van de dataset. Bereken met Python de relatieve frequentie van *C*, door gebruik te maken van de absolute frequentie van *C*. Print deze relatieve frequentie nadien op het scherm.

<div class="alert alert-block alert-danger"> 
Tip: De absolute frequentie van C en de grootte van de dataset hadden we eerder al in variabelen opgeslaan. Het is een goede gewoonte om tijdens het programmeren zo veel mogelijk met eerder gedefiniëerde variabelen te werken. Zo worden berekeningen niet onnodig meermaals uitgevoerd en vermijdt je fouten.
</div>

0.3333333333333333

In onderdeel 1.2 zagen we dat we ook operaties op een hele kolom konden uitvoeren. Deel de absolute frequentietabel van *kolom1* door de grootte van de dataset en print de resulterende tabel op het scherm.

A    0.555556
C    0.333333
B    0.111111
Name: Kolom 1, dtype: float64

Het resultaat is een **relatieve frequentietabel**. Die bevat de relatieve frequenties van alle elementen.

Wat merk je over de som van de elementen in de relatieve frequentietabel?

Antwoord: De som van de relatieve frequenties van alle elementen is 1.

Net als bij een absolute frequentietabel kunnen we bij deze nieuwe tabel ook de relatieve frequentie opvragen door de naam van een element als sleutel te gebruiken. Print via deze methode de relatieve frequentie van *C* een tweede keer op het scherm.

0.3333333333333333

Een relatieve frequentie komt steeds uit het interval [0,1]. Voer een operatie uit op de relatieve frequentietabel van *kolom1* zodat de relatieve frequenties als percentages zichtbaar zijn.

A    55.555556
C    33.333333
B    11.111111
Name: Kolom 1, dtype: float64

<div>
    <font color=#690027 markdown="1">
        <h2>2. Toepassing: Ruimtemissies</h2> 
    </font>
</div>

<div>
    <font color=#690027 markdown="1">
        <h3>2.1. Absolute frequentie</h3> 
    </font>
</div>

### Voorbeeld 2.1

Als volgende voorbeeld gebruiken we in dit notebook de dataset *Space Mission Launches*. Deze dataset is gegeven in een csv-bestand op de locatie `data/mission_launches.csv`. Met de methode `read_csv` uit de *Pandas* module kunnen we de dataset inlezen en opslaan als de variabele `ruimtemissies`.

In [15]:
ruimtemissies = pd.read_csv("data/mission_launches.csv")    # De dataset uitlezen
ruimtemissies                                               # De dataset op het scherm tonen

Unnamed: 0.1  Unnamed: 0 Organisation  \
0                0           0       SpaceX   
1                1           1         CASC   
2                2           2       SpaceX   
3                3           3    Roscosmos   
4                4           4          ULA   
...            ...         ...          ...   
4319          4319        4319      US Navy   
4320          4320        4320         AMBA   
4321          4321        4321      US Navy   
4322          4322        4322    RVSN USSR   
4323          4323        4323    RVSN USSR   

                                               Location  \
0            LC-39A, Kennedy Space Center, Florida, USA   
1     Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2                         Pad A, Boca Chica, Texas, USA   
3          Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4              SLC-41, Cape Canaveral AFS, Florida, USA   
...                                                 ...   
4319           LC-18A, Cape Canaveral AFS, Florida, USA   
4320           LC-26A, Cape Canaveral AFS, Florida, USA   
4321           LC-18A, Cape Canaveral AFS, Florida, USA   
4322          Site 1/5, Baikonur Cosmodrome, Kazakhstan   
4323          Site 1/5, Baikonur Cosmodrome, Kazakhstan   

                            Date  \
0     Fri Aug 07, 2020 05:12 UTC   
1     Thu Aug 06, 2020 04:01 UTC   
2     Tue Aug 04, 2020 23:57 UTC   
3     Thu Jul 30, 2020 21:25 UTC   
4     Thu Jul 30, 2020 11:50 UTC   
...                          ...   
4319  Wed Feb 05, 1958 07:33 UTC   
4320  Sat Feb 01, 1958 03:48 UTC   
4321  Fri Dec 06, 1957 16:44 UTC   
4322  Sun Nov 03, 1957 02:30 UTC   
4323  Fri Oct 04, 1957 19:28 UTC   

                                            Detail  Rocket_Status  Price  \
0     Falcon 9 Block 5 | Starlink V1 L9 & BlackSky   StatusActive   50.0   
1              Long March 2D | Gaofen-9 04 & Q-SAT   StatusActive  29.75   
2               Starship Prototype | 150 Meter Hop   StatusActive    NaN   
3     Proton-M/Briz-M | Ekspress-80 & Ekspress-103   StatusActive   65.0   
4                       Atlas V 541 | Perseverance   StatusActive  145.0   
...                                            ...            ...    ...   
4319                     Vanguard | Vanguard TV3BU  StatusRetired    NaN   
4320                           Juno I | Explorer 1  StatusRetired    NaN   
4321                       Vanguard | Vanguard TV3  StatusRetired    NaN   
4322                    Sputnik 8K71PS | Sputnik-2  StatusRetired    NaN   
4323                    Sputnik 8K71PS | Sputnik-1  StatusRetired    NaN   

     Mission_Status  
0           Success  
1           Success  
2           Success  
3           Success  
4           Success  
...             ...  
4319        Failure  
4320        Success  
4321        Failure  
4322        Success  
4323        Success  

[4324 rows x 9 columns]

<div class="alert alert-box alert-success">
Als we de dataset bekijken ziet deze er op het eerste zicht vrij ingewikkeld uit. De dataset is afkomstig van <b>Kaggle</b>. Dat is platform dat onder andere datasets gratis beschikbaar stelt. Je kan de bron van deze data zelf checken via <a href="https://www.kaggle.com/datasets/sefercanapaydn/mission-launches">https://www.kaggle.com/datasets/sefercanapaydn/mission-launches</a>. Daar vind je meer info over waar de data juist vandaan komt, en wat elke kolom wil zeggen.    
</div> 

De dataset bevat 4324 rijen die elk gegevens bevatten over 1 ruimtemissie. Elk soort gegeven wordt in een aparte kolom weergegeven. We willen voor elke organisatie onderzoeken hoeveel ruimtemissies zij al hebben uitgevoerd.

We hebben hiervoor enkel de kolom `Organisation`. Dus we slaan die kolom alvast op in een aparte variabele.

In [16]:
organisaties = ruimtemissies["Organisation"]  # De kolom Organisation uit de dataset ruimtemissies uitlezen en oplaan in de variabele organisaties
organisaties  # De kolom Organisation op het scherm printen

0          SpaceX
1            CASC
2          SpaceX
3       Roscosmos
4             ULA
          ...    
4319      US Navy
4320         AMBA
4321      US Navy
4322    RVSN USSR
4323    RVSN USSR
Name: Organisation, Length: 4324, dtype: object

Om de absolute frequentietabel op te stellen moeten we tellen hoeveel keer elke organisatie in de kolom *Organisation* voorkomt. 4324 waarden tellen zou voor een mens enkele uren werk zijn. Voor een computer duurt dat enkele milliseconden. We kunnen hiervoor de `value_counts` functie gebruiken.

In [17]:
absolute_frequentietabel_organisaties = organisaties.value_counts()  # Stel de absolute frequentietabel van organisaties op en sla ze op in de variabele absolute_frequentietabel_organisaties
absolute_frequentietabel_organisaties  # Print de absolute frequentietabel op het scherm

RVSN USSR           1777
Arianespace          279
CASC                 251
General Dynamics     251
NASA                 203
VKS RF               201
US Air Force         161
ULA                  140
Boeing               136
Martin Marietta      114
SpaceX               100
MHI                   84
Northrop              83
Lockheed              79
ISRO                  76
Roscosmos             55
ILS                   46
Sea Launch            36
ISAS                  30
Kosmotras             22
US Navy               17
ISA                   13
Rocket Lab            13
Eurockot              13
ESA                   13
Blue Origin           12
IAI                   11
ExPace                10
ASI                    9
CNES                   8
AMBA                   8
MITT                   7
JAXA                   7
Land Launch            7
UT                     5
KCST                   5
CASIC                  5
Exos                   4
CECLES                 4
Arm??e de l'Air        4


Met absolute frequentietabel ter beschikking kunnen we uit de dataset, die op het eerste zicht ingewikkeld leek, al een hoop nuttige informatie afleiden.

### Opdracht 2.1
Welke organisatie heeft al de meeste missies volbracht?

Antwoord:

### Opdracht 2.2
Hoeveel missies heeft Roscosmos al volbracht?

Antwoord:

### Voorbeeld 2.2

Uiteraard kunnen we dit aflezen uit de absolute frequentietabel. Maar we kunnen dit ook opvragen in code door een gegeven als sleutel van de kolom te gebruiken. Zo kunnen we ook berekeningen uitvoeren met de absolute frequentie van een gegeven.

Onderstaand codefragment slaat de absolute frequentie van het aantal ruimtemissies door Boeing op in een variabele

In [18]:
absolute_frequentie_Boeing = absolute_frequentietabel_organisaties["Boeing"]  # Sla de absulute frequentie van de organisatie "Boeing" op in de variabele absolute_frequentie_Boeing
absolute_frequentie_Boeing                                                    # Print de absulute frequentie van "Boeing" op het scherm

136

### Opdracht 2.3
Bereken de absolute frequentie van de waarde SpaceX en sla deze op in de variabele `absolute_frequentie_SpaceX`.  Print nadien de som van het aantal ruimtemissies dat door Boeing of SpaceX werd georganiseerd op het scherm.

236

### Opdracht 2.4
Bereken de grootte van de *mission_launches* dataset en sla zo op de variabele `ruimtemissies_grootte`.

4324

### Opdracht 2.5
We willen nu weten hoeveel missies er slagen en falen. Stel hiertoe de absolute frequentietabel op van de missiestatussen. Je kan die vinden in de kolom `Mission_Status`.

0       Success
1       Success
2       Success
3       Success
4       Success
         ...   
4319    Failure
4320    Success
4321    Failure
4322    Success
4323    Success
Name: Mission_Status, Length: 4324, dtype: object

### Opdracht 2.6
Print nu de volgende gegevens op het scherm:
* Het aantal succesvolle missies
* Het aantal volledig gefaalde missies (dus de som van het aantal missies met Failure of Prelaunch Failure)
* Het totaal aantal gefaalde missies, door gebruik te maken van het totaal aantal missies.

3879
343
445


<div>
    <font color=#690027 markdown="1">
        <h3>2.2. Relatieve frequentie</h3> 
    </font>
</div>

### Voorbeeld 2.3
We willen nu weten hoeveel procent van de ruimtemissies werden uitgevoerd door NASA. Daartoe berekenen we de verhouding tussen het aantal de missies door organistie NASA en het totale aantal missies. Die verhouding kunnen we via een deling berekenen als volgt.

In [23]:
absolute_frequentie_NASA = absolute_frequentietabel_organisaties["NASA"]   # De absolute frequentietabel van de organisaties hebben we al. Daaruit halen we de absolute frequentie van NASA
relatieve_frequentie_NASA = absolute_frequentie_NASA/ruimtemissies_grootte  # Door de absolute frequentie door de grootte van de dataset te delen krijgen we de relatieve frequentie
relatieve_frequentie_NASA  # Print de relatieve frequentie op het scherm

0.0469472710453284

Ongeveer 4,69 % van de ruimtemissies werd dus door NASA georganiseerd.

### Opdracht 2.7
Bereken het percentage missies dat door voor de lancering is gefaald (`Prelaunch Failure`).

0.09250693802035154

### Voorbeeld 2.4
We kunnen ook de volledige tabel delen door de grootte van de dataset om een relatieve frequentietabel te verkrijgen. We berekenen hier de relatieve frequentietabel van alle organisaties.

In [25]:
relatieve_frequentietabel_organisaties = absolute_frequentietabel_organisaties/ruimtemissies_grootte  # Bereken de relatieve frequentietabel van de organisaties
relatieve_frequentietabel_organisaties  # Print de relatieve frequentie op het scherm

RVSN USSR           0.410962
Arianespace         0.064524
CASC                0.058048
General Dynamics    0.058048
NASA                0.046947
VKS RF              0.046485
US Air Force        0.037234
ULA                 0.032377
Boeing              0.031452
Martin Marietta     0.026364
SpaceX              0.023127
MHI                 0.019426
Northrop            0.019195
Lockheed            0.018270
ISRO                0.017576
Roscosmos           0.012720
ILS                 0.010638
Sea Launch          0.008326
ISAS                0.006938
Kosmotras           0.005088
US Navy             0.003932
ISA                 0.003006
Rocket Lab          0.003006
Eurockot            0.003006
ESA                 0.003006
Blue Origin         0.002775
IAI                 0.002544
ExPace              0.002313
ASI                 0.002081
CNES                0.001850
AMBA                0.001850
MITT                0.001619
JAXA                0.001619
Land Launch         0.001619
UT            

Uit deze tabel kunnen we ook de relatieve frequentie van een gegeven halen op dezelfde manier dan de absolute frequentietabel.

In [26]:
relatieve_frequentietabel_organisaties["NASA"]

0.0469472710453284

Dit komt inderdaad overeen met de relatieve frequentie uit voorbeeld 3.1.

### Opdracht 2.8
Genereer de relatieve frequentietabel van de status van elke missie. Print ook het percentage missies dat voor de lancering is gefaald door enkel gebruik te maken van deze tabel.

Success              0.897086
Failure              0.078400
Partial Failure      0.023589
Prelaunch Failure    0.000925
Name: Mission_Status, dtype: float64

0.0009250693802035153


<div>
    <font color=#690027 markdown="1">
        <h2>3. Oefening: Video games sales</h2> 
    </font>
</div>

Volbreng de onderstaande opdrachten gegeven een nieuwe dataset met bron https://www.kaggle.com/datasets/gregorut/videogamesales. De dataset is als csv-bestand terug te vinden in bij `data/video_games_sales.csv`

<div class="alert alert-box alert-success">
Je zal tijdens het maken van de opgaven merken dat de dateset al even niet meer is geupdated. De meest recente data is afkomstig uit 2017.
</div>

<div class="alert alert-block alert-danger"> 
Print eerst de dataset op het scherm en surf eens naar de bron van de data. Deze bevat uitleg over wat elke kolom wil zeggen. Dit helpt om een overzicht te houden tijdens de oefening. Nadien werk je per opdracht stap per stap zoals in de voorgaande paragrafen. De hele opgave in één keer proberen volbrengen verhoogt de kans op fouten.
</div>

In [28]:
videogames = pd.read_csv("data/video_games_sales.csv")
videogames

rank                                              name platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         year         genre   publisher  na_sales  eu_sales  jp_sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       other_sales  global_sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

### Opdracht 3.1
Hoeveel videogames uit de lijst werden uitgebracht in 2015? Print de juiste waarde op het scherm.

614

### Opdracht 3.2
Genereer een relatieve frequentietabel voor de genres van elk videospel.

### Opdracht 3.3

Geef het percentage videogames uit de dataset die door Nintendo werden uitgebracht? Print de juiste waarde op het scherm in procent.

4.2354500542234

### Opdracht 3.4

Geef het percentage videogames uit de lijst die zijn uitgebracht op voor de playstation. Dus de videogames met platform PS, PS2, PS3, PS4 of PSP. Print de juiste waarde in procent op het scherm.

0.37564766839378244

### Referentielijst

Space Mission Launches. https://www.kaggle.com/datasets/sefercanapaydn/mission-launches

Video Game Sales. https://www.kaggle.com/datasets/gregorut/videogamesales

<img src="images/cclic.png" alt="Banner" align="left" width="100"/><br><br>

Notebook python in wiskunde - Statistiek, van D. De Bolster, F. wyffels & N. Gesquière in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>.